src/main/java/org/elasticsearch/search/aggregations/reducers/bucketmetrics/BucketMetricsReducer.java

```diff
@@ -0,0 +1,99 @@
+
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.search.aggregations.Aggregation;
+import org.elasticsearch.search.aggregations.Aggregations;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.InternalAggregation.ReduceContext;
+import org.elasticsearch.search.aggregations.InternalMultiBucketAggregation;
+import org.elasticsearch.search.aggregations.bucket.MultiBucketsAggregation.Bucket;
+import org.elasticsearch.search.aggregations.reducers.BucketHelpers;
+import org.elasticsearch.search.aggregations.reducers.BucketHelpers.GapPolicy;
+import org.elasticsearch.search.aggregations.reducers.Reducer;
+import org.elasticsearch.search.aggregations.reducers.SiblingReducer;
+import org.elasticsearch.search.aggregations.support.AggregationPath;
+import org.elasticsearch.search.aggregations.support.format.ValueFormatter;
+import org.elasticsearch.search.aggregations.support.format.ValueFormatterStreams;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public abstract class BucketMetricsReducer extends SiblingReducer {
+
+    protected ValueFormatter formatter;
+    protected GapPolicy gapPolicy;
+
+    public BucketMetricsReducer() {
+        super();
+    }
+
+    protected BucketMetricsReducer(String name, String[] bucketsPaths, GapPolicy gapPolicy, @Nullable ValueFormatter formatter,
+            Map<String, Object> metaData) {
+        super(name, bucketsPaths, metaData);
+        this.gapPolicy = gapPolicy;
+        this.formatter = formatter;
+    }
+
+    public final InternalAggregation doReduce(Aggregations aggregations, ReduceContext context) {
+        preCollection();
+        List<String> bucketsPath = AggregationPath.parse(bucketsPaths()[0]).getPathElementsAsStringList();
+        for (Aggregation aggregation : aggregations) {
+            if (aggregation.getName().equals(bucketsPath.get(0))) {
+                bucketsPath = bucketsPath.subList(1, bucketsPath.size());
+                InternalMultiBucketAggregation multiBucketsAgg = (InternalMultiBucketAggregation) aggregation;
+                List<? extends Bucket> buckets = multiBucketsAgg.getBuckets();
+                for (int i = 0; i < buckets.size(); i++) {
+                    Bucket bucket = buckets.get(i);
+                    Double bucketValue = BucketHelpers.resolveBucketValue(multiBucketsAgg, bucket, bucketsPath, gapPolicy);
+                    if (bucketValue != null && !Double.isNaN(bucketValue)) {
+                        collectBucketValue(bucket.getKeyAsString(), bucketValue);
+                    }
+                }
+            }
+        }
+        return buildAggregation(Collections.EMPTY_LIST, metaData());
+    }
+
+    protected void preCollection() {
```

Comment:
Oh sorry I was confused because I thought about preCollection on aggregators which has a totally different meaning. Can we maybe rename it? maybe preReduce()?